derived `Clone` impl does the wrong thing in the presence of auto-deref

**Summary**
`autoimpl(Clone)` produces code that doesn't compile when the type contains references to types that impl `Clone`. `derive` handles this correctly using [fully-qualified function calls](https://doc.rust-lang.org/reference/expressions/call-expr.html#disambiguating-function-calls).

**To Reproduce**
```rust
#[impl_tools::autoimpl(Clone)]
struct ContainsRefImplTools<'a, T: Clone> {
    a: &'a T,
}

#[derive(Clone)]
struct ContainsRefDerive<'a, T: Clone> {
    a: &'a T,
}
```

**Expansion**
```rust
struct ContainsRefImplTools<'a, T: Clone> {
    a: &'a T,
}
#[automatically_derived]
impl<'a, T: Clone> ::core::clone::Clone for ContainsRefImplTools<'a, T> {
    fn clone(&self) -> Self {
        ContainsRefImplTools {
            a: self.a.clone(),
        }
    }
}
struct ContainsRefDerive<'a, T: Clone> {
    a: &'a T,
}
#[automatically_derived]
impl<'a, T: ::core::clone::Clone + Clone> ::core::clone::Clone
for ContainsRefDerive<'a, T> {
    #[inline]
    fn clone(&self) -> ContainsRefDerive<'a, T> {
        ContainsRefDerive {
            a: ::core::clone::Clone::clone(&self.a),
        }
    }
}
```

**Additional information**
```
error[E0308]: mismatched types
   --> src/lib.rs:762:5
    |
762 |     #[impl_tools::autoimpl(Clone)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&T`, found type parameter `T`
763 |     struct ContainsRefImplTools<'a, T: Clone> {
    |                                     - this type parameter
    |
    = note:   expected reference `&T`
            found type parameter `T`
    = note: this error originates in the attribute macro `impl_tools::autoimpl` (in Nightly builds, run with -Z macro-backtrace for more info)
help: try removing the method call
    |
762 |     #[impl_tools::autoimpl(Clone)]
    |
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.9.1] — 2023-09-07
+
+-   Fix clone for fields which auto-deref (issue #34)
+
 ## [0.9.0] — 2023-06-28
 
 -   Update to syn v2.0.0

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "impl-tools"
-version = "0.9.0"
+version = "0.9.1"
 authors = ["Diggory Hardy <git@dhardy.name>"]
 edition = "2021"
 license = "MIT/Apache-2.0"
@@ -22,7 +22,7 @@ default-features = false
 version = "2.0.0"
 
 [dependencies.impl-tools-lib]
-version = "0.9.0"
+version = "0.9.1"
 path = "lib"
 
 [dev-dependencies]

--- a/lib/Cargo.toml
+++ b/lib/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "impl-tools-lib"
-version = "0.9.0"
+version = "0.9.1"
 authors = ["Diggory Hardy <git@dhardy.name>"]
 edition = "2021"
 license = "MIT/Apache-2.0"

--- a/lib/src/autoimpl/impl_misc.rs
+++ b/lib/src/autoimpl/impl_misc.rs
@@ -34,7 +34,7 @@ impl ImplTrait for ImplClone {
                     let idents = fields.named.iter().map(|f| f.ident.as_ref().unwrap());
                     let clones = fields.named.iter().map(|f| {
                         let ident = f.ident.as_ref().unwrap();
-                        quote! { #ident: #ident.clone() }
+                        quote! { #ident: ::core::clone::Clone::clone(&#ident) }
                     });
                     quote! { #tag { #(#idents),* } => #tag { #(#clones),* }, }
                 }
@@ -45,7 +45,7 @@ impl ImplTrait for ImplClone {
                     for i in 0..len {
                         let ident = idfmt.make_call_site(format_args!("_{i}"));
                         bindings.push(quote! { ref #ident });
-                        items.push(quote! { #ident.clone() });
+                        items.push(quote! { ::core::clone::Clone::clone(&#ident) });
                     }
                     quote! { #tag ( #(#bindings),* ) => #tag ( #(#items),* ), }
                 }
@@ -72,7 +72,9 @@ impl ImplTrait for ImplClone {
                     if args.ignore_named(ident) {
                         toks.append_all(quote! { #ident: Default::default(), });
                     } else {
-                        toks.append_all(quote! { #ident: self.#ident.clone(), });
+                        toks.append_all(
+                            quote! { #ident: ::core::clone::Clone::clone(&self.#ident), },
+                        );
                     }
                 }
                 quote! { #type_ident { #toks } }
@@ -84,7 +86,7 @@ impl ImplTrait for ImplClone {
                     if args.ignore_unnamed(&index) {
                         toks.append_all(quote! { Default::default(), });
                     } else {
-                        toks.append_all(quote! { self.#index.clone(), });
+                        toks.append_all(quote! { ::core::clone::Clone::clone(&self.#index), });
                     }
                 }
                 quote! { #type_ident ( #toks ) }

--- a/tests/autoimpl.rs
+++ b/tests/autoimpl.rs
@@ -34,6 +34,19 @@ fn wrapper() {
     assert_eq!(format!("{:?}", Wrapper((), ())), "Wrapper((), _)");
 }
 
+#[autoimpl(Clone)]
+struct ContainsRef<'a, T: Clone> {
+    a: &'a T,
+}
+
+#[test]
+fn contains_ref() {
+    let a = 1;
+    let b = ContainsRef { a: &a };
+    let c = b.clone();
+    let _ = c.a;
+}
+
 #[autoimpl(Clone, Default where A: trait, B: trait)]
 #[autoimpl(Debug ignore self.c where A: Debug)]
 struct X<A, B: Debug, C> {

